db.go

```diff
@@ -127,48 +150,93 @@ func (x *dummyAuthenticator) ResetPasswordFinish(identity string, token string,
 	}
 }
 
-func (x *dummyAuthenticator) CreateIdentity(identity, password string) error {
+func (x *dummyUserStore) CreateIdentity(identity, password string) error {
 	x.usersLock.Lock()
 	defer x.usersLock.Unlock()
-	if _, exists := x.users[CanonicalizeIdentity(identity)]; !exists {
-		x.users[CanonicalizeIdentity(identity)] = &dummyUser{identity, password, ""}
+	user := getUser(x.users, identity)
+	if user == nil {
+		userId := generateUserId(x.users)
+		x.users[userId] = &dummyUser{userId, identity, password, ""}
 		return nil
 	} else {
 		return ErrIdentityExists
 	}
 }
 
-func (x *dummyAuthenticator) RenameIdentity(oldIdent, newIdent string) error {
+func (x *dummyUserStore) RenameIdentity(oldIdent, newIdent string) error {
 	x.usersLock.Lock()
 	defer x.usersLock.Unlock()
 
-	if _, exists := x.users[CanonicalizeIdentity(newIdent)]; exists {
-		return ErrIdentityExists
-	}
-
-	oldIdent = CanonicalizeIdentity(oldIdent)
 	newKey := CanonicalizeIdentity(newIdent)
+	oldIdent = CanonicalizeIdentity(oldIdent)
+	newUser := getUser(x.users, newKey)
+	if newUser == nil {
+		oldUser := getUser(x.users, oldIdent)
 
-	if user, exists := x.users[oldIdent]; exists {
-		delete(x.users, oldIdent)
-		x.users[newKey] = &dummyUser{newIdent, user.password, ""}
-		return nil
+		if oldUser != nil {
+			x.users[oldUser.userId].name = newIdent
+			return nil
+		} else {
+			return ErrIdentityAuthNotFound
+		}
 	} else {
-		return ErrIdentityAuthNotFound
+		return ErrIdentityExists
 	}
 }
 
-func (x *dummyAuthenticator) GetIdentities() ([]string, error) {
+func (x *dummyUserStore) GetIdentities() ([]string, error) {
 	x.usersLock.RLock()
 	list := []string{}
-	for identity, _ := range x.users {
-		list = append(list, identity)
+	for _, v := range x.users {
+		list = append(list, v.name)
 	}
 	x.usersLock.RUnlock()
 	return list, nil
 }
 
-func (x *dummyAuthenticator) Close() {
+func (x *dummyUserStore) GetIdentityFromUserId(userId UserId) (string, error) {
+	x.usersLock.RLock()
+	defer x.usersLock.RUnlock()
+
+	for _, v := range x.users {
+		if v.userId == userId {
+			return v.name, nil
+		}
+	}
+
+	return "", ErrIdentityAuthNotFound
+}
+
+func (x *dummyUserStore) GetUserIdFromIdentity(identity string) (UserId, error) {
+	x.usersLock.RLock()
+	defer x.usersLock.RUnlock()
+
+	if CanonicalizeIdentity(identity) == "" {
+		return -1, ErrIdentityEmpty
+	}
+
+	for _, v := range x.users {
+		if v.name == identity {
+			return v.userId, nil
+		}
+	}
+	return -1, ErrIdentityAuthNotFound
+}
+
+func (x *dummyUserStore) Close() {
+}
+
+func getUser(users map[UserId]*dummyUser, identity string) *dummyUser {
```

Comment:
I'm pretty sure getUser and generateUserId are only used from methods of dummyUserStore, so they might as well be member functions.
Also - does it not complicate unit tests to have UserIds recycled? I think it would be more consistent with real-world database behaviour if UserIds were generated from a 64-bit integer that just always increments by one, and is only reset when the dummyUserStore object is discarded.